components/gfx/display_list/mod.rs

```diff
@@ -521,26 +336,201 @@ impl DisplayList {
+            let inv_transform = stacking_context.transform.inverse().unwrap();
+            let frac_point = inv_transform.transform_point(&Point2D::new(point.x.to_f32_px(),
+                                                                         point.y.to_f32_px()));
+            Point2D::new(Au::from_f32_px(frac_point.x), Au::from_f32_px(frac_point.y))
+        };
+
+        // Adjust the translated point to account for the scroll offset if
+        // necessary. This can only happen when WebRender is in use.
+        //
+        // We don't perform this adjustment on the root stacking context because
+        // the DOM-side code has already translated the point for us (e.g. in
+        // `Window::hit_test_query()`) by now.
+        if !is_fixed && stacking_context.id != StackingContextId::root() {
+            if let Some(scroll_offset) = scroll_offsets.get(&stacking_context.id) {
+                translated_point.x -= Au::from_f32_px(scroll_offset.x);
+                translated_point.y -= Au::from_f32_px(scroll_offset.y);
+            }
+        }
+
+        self.hit_test_contents(traversal, &translated_point, client_point, scroll_offsets, result);
+    }
+
+    pub fn print(&self) {
+        let mut print_tree = PrintTree::new("Display List".to_owned());
+        self.print_with_tree(&mut print_tree);
+    }
+
+    pub fn print_with_tree(&self, print_tree: &mut PrintTree) {
+        print_tree.new_level("Items".to_owned());
+        for item in &self.list {
+            print_tree.add_item(format!("{:?} StackingContext: {:?}",
+                                        item,
+                                        item.base().stacking_context_id));
+        }
+        print_tree.end_level();
+    }
+}
+
+pub struct DisplayListTraversal<'a> {
+    pub display_list: &'a DisplayList,
+    pub next_item_index: usize,
+    pub first_item_index: usize,
+    pub last_item_index: usize,
+}
+
+impl<'a> DisplayListTraversal<'a> {
+    pub fn new(display_list: &'a DisplayList) -> DisplayListTraversal {
+        DisplayListTraversal {
+            display_list: display_list,
+            next_item_index: 0,
+            first_item_index: 0,
+            last_item_index: display_list.list.len(),
+        }
+    }
+
+    pub fn new_partial(display_list: &'a DisplayList,
+                       stacking_context_id: StackingContextId,
+                       start: usize,
+                       end: usize)
+                       -> DisplayListTraversal {
+        assert!(display_list.list.len() > start);
+        assert!(display_list.list.len() > end);
+
+        let mut stacking_context_start = start;
+        for x in (0..start).rev()  {
+            if let DisplayItem::PushStackingContextClass(ref item) = display_list.list[x] {
+                if item.stacking_context.id == stacking_context_id {
+                    stacking_context_start = x;
+                    break;
+                }
+            }
+        }
```

Comment:
Can you use `display_list.list.iter().rposition(...).unwrap_or(stacking_context_start)`?